docs/proposals/gpu-support.md

```diff
@@ -0,0 +1,263 @@
+- some vendors require fairly tight coupling between the kernel driver
+controlling the GPU and the libraries/applications that access the hardware
+- it adds more resource types (whole GPUs, GPU cores, GPU memory)
+- it can introduce new security pitfalls
+- for systems with multiple GPUs, affinity matters, similarly to NUMA
+considerations for CPUs
+- running GPU code in containers is still a relatively novel idea
+
+## Detailed discussion
+
+Currently, this document is mostly focused on the basic use case: run GPU code
+on AWS `g2.2xlarge` EC2 machine instances using Docker. It constitutes a narrow
+enough scenario that it does not require large amounts of generic code yet. GCE
+doesn't support GPUs at all; bare metal systems throw a lot of extra variables
+into the mix.
+
+Later sections will outline future work to support a broader set of hardware,
+environments and container runtimes.
+
+### Inventory
+
+Before any scheduling can occur, we need to know what's available out there. For
+v0, we can punt on having automatic detection of resources. Instead, we rely on
+node labels such as `alpha.kubernetes.io/gpu-units`,
+`alpha.kubernetes.io/gpu-cores` and `alpha.kubernetes.io/gpu-memory` to provide
+the information. Administrators will populate the labels through a method of
+their choice. They could run, periodically and from a central location, a script
+that looks for machines with the `beta.kubernetes.io/instance-type=X` label and
+tag those accordingly. Otherwise, they could run a DaemonSet that performs the
+equivalent from each of the nodes (relatively expensive). Or, perhaps, they can
+reuse and repurpose existing automation and so on.
+
+### Scheduling
+
+In the long run, GPUs should be visible as first-class resources, the way CPUs
+are. That requires extensive changes to the scheduling, after having accurate
+inventory; even more so when nodes can have more than one device. It's probable
+that GPUs will exacerbate the need for [a rescheduler](rescheduler.md) or pod
+priorities, especially if the nodes in a cluster are not homogeneous. Consider
+these two cases:
+
+> Only half of the machines have a GPU and they're all busy with other
+workloads. The other half of the cluster is doing very little work. A GPU
+workload comes, but it can't schedule, because the devices are sitting idle on
+nodes that are running something else and the nodes with little load lack the
+hardware.
+
+> Some or all the machines have two graphic cards each. A number of jobs get
+scheduled, requesting one device per pod. The scheduler puts them all on
+different machines, spreading the load, perhaps by design. Then a new job comes
+in, requiring two devices per pod, but it can't schedule anywhere, because all
+we can find, at most, is one unused device per node.
+
+For v0, we'll use [a scheduler extender](../design/scheduler_extender.md). An
+extender runs outside of `kube-scheduler` and is consulted before picking a
```

Comment:
There are important details that are not obvious from the extender design document, e.g. the fact that extenders are run last and that they themselves are run in order.